# 9번

### 문제
어떤 자연수 n이 있을 때, d(n)을 n의 각 자릿수 숫자들과 n 자신을 더한 숫자라고 정의하자.   
n을 d(n)의 제네레이터(generator)라고 한다.  
예를 들어 d(91) = 9 + 1 + 91 = 101인 경우 91은 101의 제네레이터이다.  
어떤 숫자들은 하나 이상의 제네레이터를 가지고 있는데, 101의 제네레이터는 91 뿐 아니라 100도 있다.  
그런데 반대로, 제네레이터가 없는 숫자들도 있으며, 이런 숫자를 인도의 수학자 Kaprekar가 셀프 넘버(self-number)라 이름 붙였다.  

예를 들어 1,3,5,7,9,20,31 은 셀프 넘버 들이다.  

1 이상이고 5000 보다 작은 모든 셀프 넘버들의 합을 구하라.  

### 과제 1.
1부터 30까지의 자연수 중에서, 제너레이터가 있는 수의 합을 구하시오.  
출력 : 359

In [9]:
def d(n):
    for i in str(n): #n을 문자열로 변경
        n+= int(i) # n을 정수로 변경
    return n

a= set(list(range(1,30))) # 임의로 n설정
b= set() # 제너레이터에 대한 집합

for i in a:
    if d(i)<= 30 : # 30이하로 범위 설정
        b.add(d(i)) 

print(sum(b))

420


d(n)의 제너레이터는 n이다. 

d(1)=0+1+1=2  
d(2)=0+2+2=4  
d(3)=0+3+3=6  
d(4)=0+4+4=8  
d(5)=0+5+5=10  
d(6)=0+6+6=12  
d(7)=0+7+7=14   
d(8)=0+8+8=16  
d(9)=0+9+9=18  
d(10)=1+0+10=11  
d(11)=1+1+11=13  
d(12)=1+2+12=15  
d(13)=1+3+13=17  
d(14)=1+4+14=19  
d(15)=1+5+15=21  
d(16)=1+6+16=23  
d(17)=1+7+17=25  
d(18)=1+8+18=27  
d(19)=1+9+19=29  
d(20)=2+0+20=22  
d(21)=2+1+21=24  
d(22)=2+2+22=26  
d(23)=2+3+23=28  
d(24)=2+4+24=30  -> 여기까지 해당
d(25)=2+5+25=32  
d(26)=2+6+26=34  
d(27)=2+7+27=36  
d(28)=2+8+28=38  
d(29)=2+9+29=40  
d(30)=3+0+30=33  
  
- 1부터 30까지의 자연수 중에서 제너레이터가 있는 수 :  
2, 4, 6, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 22, 23, 24, 27, 26, 27, 28, 29, 30  


- 이 수들의 총합은 : 420